geograph/geograph.py

```diff
@@ -591,41 +637,82 @@ def add_habitat(
         hgraph.remove_nodes_from(invalid_idx)
-
+        # get list of barrier node indices (iloc)
+        barrier_indices = set(np.where(np.isin(self.class_label, barrier_classes))[0])
         for node in tqdm(
             hgraph.nodes, desc="Generating habitat graph", total=len(hgraph)
         ):
             polygon = polygons[node]
+            repr_point = polygon.representative_point()
             if max_travel_distance > 0:
                 buff_poly_bounds = buff_polygons[node].bounds
-                buff_poly = prep(buff_polygons[node])
+                buff_poly = buff_polygons[node]
             else:
                 buff_poly_bounds = polygon.bounds
-                buff_poly = prep(polygon)
+                buff_poly = polygon
+            if len(barrier_classes) == 0:
+                buff_poly = prep(buff_poly)
             # Query rtree for all polygons within `max_travel_distance` of the original
-            for nbr in self.rtree.intersection(buff_poly_bounds):
-                # Necessary to correct for the rtree returning iloc indexes
-                nbr = idx_dict[nbr]
-                # If a node is not a habitat class node, don't add the edge
-                if nbr == node or nbr in invalid_idx:
+            nbrs = set(self.rtree.intersection(buff_poly_bounds))
+            barrier_nbrs = {idx_dict[nbr] for nbr in barrier_indices.intersection(nbrs)}
+            barriers_exist = len(barrier_nbrs) > 0
+            barrier_poly = self.df["geometry"].loc[barrier_nbrs].unary_union
+            # Necessary to correct for the rtree returning iloc indexes
+            nbrs = {idx_dict[nbr] for nbr in nbrs}
+            for nbr in nbrs:
+                # If a node is not a habitat class node or is a barrier class node,
+                # don't add the edge
+                if nbr == node or nbr in barrier_nbrs or nbr in invalid_idx:
                     continue
                 # Otherwise add the edge with distance attribute
                 nbr_polygon = polygons[nbr]
                 if not hgraph.has_edge(node, nbr) and buff_poly.intersects(nbr_polygon):
-                    if add_distance:
-                        if max_travel_distance == 0:
-                            dist = 0.0
+                    add_edge = True
+                    if barriers_exist:
+                        # Here we loop through all barrier_nbrs and calculate the
+                        # set difference between the buffered polygon and the
+                        # barrier polygon. If this results in multiple polygons,
+                        # then the barrier polygon fully cuts the buffered polygon.
+                        #
+                        # Therefore we find the largest polygon in the result,
+                        # which will contain the original node polygon, and check
+                        # if it intersects the neighbour polygon we're trying to
+                        # reach. If it does not, there is no path.
```

Comment:
Looks good, thank you for going through the comments in detail @herbiebradley . Here one last suggestion to update the inline documentation with what happens in the new implementation of the method.

```suggestion
                        # We find the resulting buffered polygon fragment that contains 
                        # the original node polygon, by selecting the polygon that 
                        # contains a representative point sampled from the original 
                        # polygon. 
                        # We then check if that polygon fragment intersects the 
                        # neighbour polygon we're trying to reach. 
                        # If it does not, there is no path.
```